File: scripts/update_leaderboard.py
import os
from supabase import create_client, Client
from datetime import datetime, timedelta
import logging

logging.basicConfig(level=logging.INFO)

# Initialize Supabase client
SUPABASE_URL = os.getenv("SUPABASE_URL")
SUPABASE_KEY = os.getenv("SUPABASE_API_KEY")
supabase: Client = create_client(SUPABASE_URL, SUPABASE_KEY)

def update_leaderboard():
    """Checks for users who haven't played yesterday and updates their leaderboard stats."""
    if not SUPABASE_URL or not SUPABASE_KEY:
        raise ValueError("Supabase URL or API key is not set")

    logging.info("Supabase client initialized")

    try:
        today = datetime.utcnow().date()
        yesterday = today - timedelta(days=1)  # Correctly refers to the previous day

        # Find users who haven't played on 'yesterday'
        response = supabase.table('leaderboard').select('userid').execute()
        inactive_users = [user['userid'] for user in response.data if user['userid'] not in (
            user['userid'] for user in supabase.table('games').select('userid').eq('game_date', yesterday).execute().data)]

        if inactive_users:
            user_ids = inactive_users

            # Update leaderboard for inactive users
            supabase.table('leaderboard').update({
                'games_played': supabase.func('games_played + 1'),
                'total_attempts': supabase.func('total_attempts + 10')
            }).in_('userid', user_ids).execute()

            logging.info("Leaderboard updated successfully.")

    except Exception as e:
        logging.error(f"Error updating leaderboard: {e}")
        raise

if __name__ == "__main__":
    update_leaderboard()

File: src/app/api/update-leaderboard.js
import { supabase } from "@/utils/supabase";

export default async function handler(req, res) {
  const { userId, attempts } = req.body;

  try {
    const { data: userData, error: userError } = await supabase
      .from("users")
      .select("username")
      .eq("user_id", userId)
      .maybeSingle();

    if (userError || !userData) {
      return res.status(500).json({ error: userError?.message || "User not found" });
    }

    const username = userData.username || "Unknown";

    const { data, error } = await supabase
      .from("leaderboard")
      .select("*")
      .eq("user_id", userId)
      .maybeSingle();

    if (error) {
      return res.status(500).json({ error: error.message });
    }

    if (data) {
      const { error: updateError } = await supabase
        .from("leaderboard")
        .update({
          total_attempts: data.total_attempts + attempts,
          games_played: data.games_played + 1,
        })
        .eq("user_id", userId);

      if (updateError) {
        return res.status(500).json({ error: updateError.message });
      }
    } else {
      const { error: insertError } = await supabase.from("leaderboard").insert([{
        user_id: userId,
        username: username,
        total_attempts: attempts,
        games_played: 1,
      }]);

      if (insertError) {
        return res.status(500).json({ error: insertError.message });
      }
    }

    const today = new Date().toISOString().slice(0, 10);

    const { data: existingGame, error: fetchError } = await supabase
      .from("games")
      .select("id")
      .eq("user_id", userId)
      .maybeSingle();

    if (fetchError) {
      return res.status(500).json({ error: fetchError.message });
    } else if (existingGame) {
      const { error: updateError } = await supabase
        .from("games")
        .update({ date: today, attempts })
        .eq("id", existingGame.id);

      if (updateError) {
        return res.status(500).json({ error: updateError.message });
      }
    } else {
      const { error: insertError } = await supabase.from("games").insert([
        {
          user_id: userId,
          date: today,
          attempts: attempts,
        },
      ]);

      if (insertError) {
        return res.status(500).json({ error: insertError.message });
      }
    }

    res.status(200).json({ message: "Leaderboard updated successfully" });
  } catch (error) {
    res.status(500).json({ error: error.message });
  }
}
